### DISCUSSION TITLE
different NMS results in train.py and val.py

### DISCUSSION THREAD

**Discussion by junghyun-avikus** [ultralytics/yolov5#7849] [Q&A]
Hi,
I am working on extracting TPs, FNs, FPs of the first image of a random batch in each epoch during the training to get a better understanding of models. So I am applying the same logic that used in val.py into the train.py. 

```python
 # Inference
out, train_out = model(im) if training else model(im, augment=augment, val=True)  # inference, loss outputs
dt[1] += time_sync() - t2

# Loss
if compute_loss:
    loss += compute_loss([x.float() for x in train_out], targets)[1]  # box, obj, cls

# NMS
targets[:, 2:] *= torch.tensor((width, height, width, height), device=device)  # to pixels
lb = [targets[targets[:, 0] == i, 1:] for i in range(nb)] if save_hybrid else []  # for autolabelling
t3 = time_sync()
out = non_max_suppression(out, conf_thres, iou_thres, labels=lb, multi_label=True, agnostic=single_cls)
```

This is the logic that is implemented in val.py, and it seems quite straight forward. It converts targets first, and then creates a list of lists where each list is labels of an image in a batch. Then, the out is a list of detections, where each detection is represented in [xyxy, conf, cls]

So, I added the same logic to the train.py in the following way.

```python
if i == 0:
    nb, _, height, width = imgs.shape
    conf_thres, iou_thres = 0.25, 0.45
    out, _ = ema.ema(imgs)
    targets = targets.to(device)

    targets[:, 2:] *= torch.tensor((width, height, width, height), device=device)  # to pixels
    lb = [targets[targets[:, 0] == i, 1:] for i in range(nb)]  # for autolabelling
    out = non_max_suppression(out, conf_thres, iou_thres, labels=lb, multi_label=True, agnostic=single_cls)
```

But in this case, its conf score seems to be messed up somehow. most of them are 1.000.

What am I missing here?

**Comment by glenn-jocher** [comment]
@junghyun-avikus TP and FP are output in val.py here:
https://github.com/ultralytics/yolov5/blob/614ef110b395eec842feac55f38805161f6ddb2e/val.py#L264-L265

**Reply by junghyun-avikus** [reply]
oh what I was trying to do was, extracting true positive / false positive bounding boxes out of predicted boxes that are above certain threshold in the training phase. Are there any code snippets that do something similar? It seems ap_per_class does extract the number of tps or fps.
